core/src/main/java/org/elasticsearch/cluster/metadata/IndexTombstone.java

```diff
@@ -0,0 +1,227 @@
+    }
+
+    /**
+     * The index UUID of the deleted index.
+     */
+    public String getIndexUUID() {
+        return indexUUID;
+    }
+
+    /**
+     * The index name of the deleted index.
+     */
+    public String getIndexName() {
+       return indexName;
+    }
+
+    /**
+     * The date in milliseconds that the index deletion event occurred,
+     * this is used for purging tombstone entries from the cluster state.
+     */
+    public long getDeleteDate() {
+        return deleteDate;
+    }
+
+    /**
+     * The cluster state version number (see {@link org.elasticsearch.cluster.ClusterState#version()})
+     * corresponding to when the index deletion event occurred.
+     */
+    public long getClusterVersion() {
+        return clusterVersion;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(indexUUID);
+        out.writeString(indexName);
+        out.writeLong(deleteDate);
+        out.writeVLong(clusterVersion);
+    }
+
+    @Override
+    public IndexTombstone readFrom(StreamInput in) throws IOException {
+        return new IndexTombstone(in);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+        if (other == null || getClass() != other.getClass()) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        IndexTombstone that = (IndexTombstone) other;
+        if (indexUUID.equals(that.indexUUID) == false) {
+            return false;
+        }
+        if (indexName.equals(that.indexName) == false) {
+            return false;
+        }
+        if (deleteDate != that.deleteDate) {
+            return false;
+        }
+        if (clusterVersion != that.clusterVersion) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(indexUUID, indexName, deleteDate, clusterVersion);
+    }
+
+    @Override
+    public String toString() {
+        return "IndexTombstone[indexUUID=" + indexUUID + ", indexName=" + indexName +
+               ", deleteDate=" + deleteDate + ", clusterVersion=" + clusterVersion + "]";
+    }
+
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+        builder.field(INDEX_UUID_KEY, indexUUID);
+        builder.field(INDEX_NAME_KEY, indexName);
+        builder.field(DELETE_DATE_KEY, deleteDate);
+        builder.field(CLUSTER_VERSION_KEY, clusterVersion);
+        builder.endObject();
+        return builder;
+    }
+
+    public static IndexTombstone fromXContent(final XContentParser parser) throws IOException {
+        return TOMBSTONE_PARSER.parse(parser, new IndexTombstone.Builder()).build();
+    }
+
+    /** A builder for building IndexTombstone objects. */
+    public static class Builder {
```

Comment:
The issue is, if we use the builder to create an `IndexTombstone`, then it needs to be accessible to the `MetaDataDeleteIndexService` where it is created.